kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,591 @@
+    end
+end
+
+
+local function stats_count(stats, name, key)
+    stats[name][key] = (stats[name][key] or 0) + 1
+end
+
+
+-- For TYPE_LAST: the DNS record from the last successful query
+local valid_types = {
+    [ TYPE_SRV ] = true,
+    [ TYPE_A ] = true,
+    [ TYPE_AAAA ] = true,
+    [ TYPE_CNAME ] = true,
+}
+
+local function insert_last_type(cache, name, qtype)
+    if valid_types[qtype] then
+        cache:set("last:" .. name, { ttl = 0 }, qtype)
+    end
+end
+
+
+local function get_last_type(cache, name)
+    return cache:get("last:" .. name)
+end
+
+
+-- insert hosts into cache
+local function init_hosts(cache, path, preferred_ip_type)
+    local hosts, err = utils.parse_hosts(path)
+    if not hosts then
+        ngx.log(ngx.WARN, "Invalid hosts file: ", err)
+        hosts = {}
+    end
+
+    if not hosts.localhost then
+        hosts.localhost = {
+          ipv4 = "127.0.0.1",
+          ipv6 = "[::1]",
+        }
+    end
+
+    local function insert_answer(name, qtype, address)
+        if not address then
+            return
+        end
+
+        local key = name .. ":" .. qtype
+        local answers = {{
+            name = name,
+            type = qtype,
+            address = address,
+            class = 1,
+            ttl = 0,
+        }}
+        cache:set(key, { ttl = 0 }, answers)
+    end
+
+    for name, address in pairs(hosts) do
+        name = name:lower()
+        if address.ipv4 then
+            insert_answer(name, TYPE_A, address.ipv4)
+            insert_last_type(cache, name, TYPE_A)
+        end
+        if address.ipv6 then
+            insert_answer(name, TYPE_AAAA, address.ipv6)
+            if not address.ipv4 or preferred_ip_type == TYPE_AAAA then
+                insert_last_type(cache, name, TYPE_AAAA)
+            end
+        end
+    end
+
+    return hosts
+end
+
+
+function _M.new(opts)
+    if not opts then
+        return nil, "no options table specified"
+    end
+
+    -- parse resolv.conf
+    local resolv, err = utils.parse_resolv_conf(opts.resolv_conf, enable_ipv6)
+    if not resolv then
+        ngx.log(ngx.WARN, "Invalid resolv.conf: ", err)
+        resolv = { options = {} }
+    end
+
+	-- init the resolver options for lua-resty-dns
+    local nameservers = opts.nameservers or resolv.nameservers
+    if not nameservers or #nameservers == 0 then
+        ngx.log(ngx.WARN, "Invalid configuration, no nameservers specified")
+    end
+
+    local r_opts = {
+        nameservers = nameservers,
+        retrans = opts.retrans or resolv.options.attempts or 5,
+        timeout = opts.timeout or resolv.options.timeout or 2000,   -- ms
+        no_random = opts.no_random or not resolv.options.rotate,
+    }
+
+    -- init the mlcache
+    local lock_timeout = r_opts.timeout / 1000 * r_opts.retrans + 1 -- s
+
+    local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
+        lru_size = opts.cache_size or 10000,
+        ipc_shm = "kong_dns_cache_ipc",
```

Comment:
I advise against using `ipc_shm` in Kong. This option is for mlcache to provide an off-the-shelf IPC mechanism for simple use-cases or ease of development, but IPC-based (i.e. message-based) inter-workers communication in high throughput OpenResty application is a dangerous pattern that has bitten us many times in the past.

See the note in mlcache docs:

> this library was built with the intent to use a better solution for inter-worker communication as soon as one emerges. In future versions of this library, if an IPC library can avoid the polling approach, so will this library. update() is only a necessary evil due to today's Nginx/OpenResty "limitations". You can however use your own IPC library by use of the opts.ipc option when creating your mlcache instance.

Instead, mlcache provides the `opts.ipc` mechanism which is not explicitly documented but should be straightforward to understand (see [mlcache.lua](https://github.com/thibaultcha/lua-resty-mlcache/blob/main/lib/resty/mlcache.lua#L243-L263)). Using the `opts.ipc.broadcast` and `opts.ipc.poll` functions, mlcache users can attach an mlcache instance to an IPC mechanism of their choice.

It is my understanding that today, Kong has better IPC mechanisms in place that messaged-based shm solutions. Thus I *strongly* advise for the DNS client to make use of this new IPC mechanism instead of relying on `ipc_shm`.

Reply:
Yeah, thanks! I found I could use https://github.com/Kong/lua-resty-events to deal with DNS cache RPC

Reply:
done using kong.worker_events instead of relying on mlcache defautl `ipc_shm`